### ISSUE TITLE
Possible enhancement to placement of outcome names and + icon in designer

### ISSUE THREAD

**Issue by tomy2105** [elsa-workflows/elsa-core#2404]
Currently the outcome name is placed at (approximately) half of the line connecting two activities.
And + icon seems to be placed almost at the end of the line near the "destination" activity.

This becomes confusing and hard to follow when line is long so you don't see outcome names and + on the same screen.

I'd suggest, if possible to keep outcome names and + icon close to the activity even when connecting line is long.

What do others think?

**Comment by sfmskywalker** [comment]
I'm thinking you will like the new designer that I'm currently working on (very early preview):

![elsa-3-designer-sneakpreview](https://user-images.githubusercontent.com/938393/139533362-f3ab79ab-967f-4bae-8cbd-8c3debda3ee7.gif)

This uses the awesome [X6](https://x6.antv.vision/en/examples/gallery) library which solves pretty much all of the issues you have ever listed ;)

**Comment by sfmskywalker** [comment]
PS. With regards to this issue's topic: outcome names + icon, the new designer solves this by using labels associated with the connections, and outcome names will probably be displayed as tooltips:

![port-tooltips](https://user-images.githubusercontent.com/938393/139533482-1e957d9d-caf0-46ef-b134-f94ec1fcbf66.gif)

**Comment by sfmskywalker** [comment]
Labels might look like this:

![image](https://user-images.githubusercontent.com/938393/139533535-4e1d4e6b-aaf7-4fc4-af24-54f872049d46.png)

**Comment by tomy2105** [comment]
Huh, new designer looks a lot like NodeRed :)....
I don't like the idea of hiding outcome names and placing them on tooltips only, makes it difficult to follow the workflow without a lot of mouse work.....

**Comment by sfmskywalker** [comment]
Regardless of displaying the outcome name next to the port or not, once connected, that outcome name would be displayed as a label on the connector (without requiring user to hover over connections):

![image](https://user-images.githubusercontent.com/938393/140304090-144df8db-2cf5-4f40-ae69-ba997022ec0e.png)

I don't think that would make the workflow too hard to follow?

**Comment by tomy2105** [comment]
Once connected yes, but during connecting still need to think and manually check which one is which.
And your new image shows same problem I have now, your outcome name seems half way between activities which in turns makes it hard to follow when activities are far apart.....

**Comment by sfmskywalker** [comment]
If activities are so far apart, then even if we display the outcome name next to the outbound port won’t solve that problem, since you would still need to bring the source activity into view.

One solution to that might be to repeat the outcome name across the connection. Then it doesn’t matter how large the distance between two activities is.

What do you think?

**Comment by tomy2105** [comment]
Repeating outcome name might work for activities far apart but will clutter if they are close....

In my, not so short experience :( using workflow tools, workflow branching is important to be able to follow (and outcome name gives you info why something branched).
And branching happens at outcome of an activity (so outcome names must be visibe, IMHO, and must be close to their activity).
It is useful to know all of the activities that "come into" one activity (but since it can be many) I'd say you should place this into a tooltip that shows when you hover over (incoming line into an activity, and I say line in singular because in one other task I said it would be useful to have connection lines merging whenever possible :)).

Other enhancements for easy following of workflow is:
- line highlighting on hover (so you clearly see where it goes)
- doubleclicking on outcome name (which is before line is merged into others) scrolls you automatically to destination activity
- in case of line merging, doubleclicking on "input" port of an activity gives you list of all outcomes of all the activities that are connected to this one and clicking on one of them scrolls you to that activity/outcome.

And yes, if you ever plan for activities with multiple input "ports" then they need a "input" name which must be placed close to that activity too as well as outcome ones.

**Comment by tomy2105** [comment]
One other thing I must warn you.
I have difficulty visualising how would 3 images (first is blue activities, second is "ports" with transparent SQL activity and Chinese (or whichever far east language that is, I sincerely apologize for not knowing) characters, third is with boxes and edge names) you presented fit together, so I apologize if I understood something wrong.

**Comment by tomy2105** [comment]
And one more comment (sorry today seems to be my short attention span day :().

Currently you have in designer automatic placement of activities and automatic connecting of activities upon adding them to the workflow.
For backward compatiblity reason I would love to see those features maintained in new release (your first image kind of suggests manual connecting of activities, and manual placement).

I don't exactly have a vision on how to achieve this though, so any ideas are welcome.

And why I'm flagging this as a "mayor" issue is this.

Elsa 1 had manual placement of activities and manual connection (if I recall correctly).
Elsa 2 has automatic placement and automatic connecting on activity addition.
Elsa 3 will have.....???

I ask because such "drastic" changes in "fundamental" paradigm how designer behaves are a bit.... well drastic and might not be welcomed by "plain" users (or the documentation writers, or product managers of products using Elsa)......

**Comment by sfmskywalker** [comment]
@tomy2105 Thanks a lot for all the feedback. You offered some really nice ideas, I like it.

> I have difficulty visualising how would 3 images

Sorry if that was confusing. The point was to highlight individual aspects that we are discussing (e.g. the [second image](https://github.com/elsa-workflows/elsa-core/issues/2404#issuecomment-955204418) demonstrates the possibility of showing port names upon hover, while the [third image](https://github.com/elsa-workflows/elsa-core/issues/2404#issuecomment-955204644) demonstrates the ability to display labels on connections. This would be the same name as the source port).

> Currently you have in designer automatic placement of activities and automatic connecting of activities upon adding them to the workflow.
> For backward compatiblity reason I would love to see those features maintained in new release (your first image kind of suggests manual connecting of activities, and manual placement).
>
> I don't exactly have a vision on how to achieve this though, so any ideas are welcome.

100% agree. The way I am planning on implementing that is by allowing you to either double-click or begin dragging an outbound port. Upon release, an "inline" activity picker appears (with searchbox) that lets you pick an activity. Upon choosing an activity, it will be added to the canvas and automatically be connected to the source activity, just like with the V2 designer.

Additionally, using e.g. SHIFT+Left-click and/or right-click to open context menu will also invoke the "inline" activity picker to add a disconnected activity (as an alternative to dragging one from the activity picker as seen in the [first image](https://github.com/elsa-workflows/elsa-core/issues/2404#issuecomment-955204018)).

Here's a small visualization taken from node-red to show what I mean by "inline" activity picker (don't let me confuse you; it does not demonstrate the "auto connect to source activity" I described earlier, just the inline picker):

![node-red-inline-picker](https://user-images.githubusercontent.com/938393/140406924-0d0094a2-a871-46dd-aa56-0e3c3fa1bd28.gif)

> Elsa 1 had manual placement of activities and manual connection (if I recall correctly).
> Elsa 2 has automatic placement and automatic connecting on activity addition.
> Elsa 3 will have.....???

- Elsa 1 had manual placement and linking of activities. This has the advantage of allowing free-form layouts which some users prefer, but obvious disadvantages are: lot's of clicking and easy to make a hot tangled plate of spaghetti. Not to mention other limitations from which Elsa 1 suffers, such as inability to zoom & pan, no undo-redo, no copy & paste. Yet, some people prefer this mode of designer.
- Elsa 2 has automatic placement and linking of activities. This has the advantage of automatic-layout preventing hot spaghetti plates, but the dagre layout algorithm sometimes makes odd layout decisions. Although we could probably introduce the ability to move activities around manually (which has been requested by some), it would require a significant amount of work. Not to mention other items from your wish list, such as the ability to use scrollbars to pan around, use selection rectangles, copy & paste, preventing odd jumps, and a whole slew of other subtle nuisances. Not saying it's impossible to fix, just that it takes a lot of work.
- Elsa 3 will have the best of both worlds: manual + automatic layout, labelled ports, copy & paste, zoom & pan, scrollbars, selection rectangles, navigation pane, connections that do a much better job at going around other activities, smaller activity displays and improved performance (noticeable when using lots of activities).

> I ask because such "drastic" changes in "fundamental" paradigm how designer behaves are a bit.... well drastic and might not be welcomed by "plain" users (or the documentation writers, or product managers of products using Elsa)......

Yeah I hear you.
Regardless of the drastic changes being introduced, I strongly feel that they are for the better based on various feedback received from different users and customers.

Here's what I collected in terms of features to do based on the above:

- [ ] Display port labels (inbound + outbound)
- [ ] Display connection labels (same text as outbound port)
- [ ] Highlight connection upon hover
- [ ] Double-click on connected outbound port takes you to destination activity
- [ ] Double-click on connected inbound port displays list of inbound activities. Clicking any one of them will take you to it
- [ ] Ability to add auto-connected activity. Can't use double-click because that takes you to destination activity if there is one, but perhaps just by start dragging and then releasing mouse button to invoke inline activity picker
- [ ] Tool button to invoke auto-layout
- [ ] Tool button to auto-layout diagram automatically whenever activities are added/removed

If there is still strong enough push-back from the community, we can always consider providing the Elsa 2 designer experience and just bring it over, as an option. Similarly, I am considering porting the V3 designer back to Elsa 2, also as an option to users who don't wish to migrate existing projects to V3 but still prefer the V3 design experience.

**Comment by tomy2105** [comment]
Thanks for good explanation, your bulleted list is definitely a nice summary and should cover it nicely.
I myself am proponent of free/manual layout (I'm control freak :)) but I definitely know a lot of people who like automatic one better, having support for both would definitely be ideal :).

As for backporting to Elsa 2, you have my strong voice for that!
Since you plan a lot of other changes for Elsa 3, it will be more difficult to "move" to Elsa 3 than to Elsa 2.x with new designer (and some of designer changes, as you've already guessed from my requests, I hardly wait to see....).

**Comment by sfmskywalker** [comment]
> And now the most annoying question an open source developer can get :(, sorry but I need to ask.
Is there any timelime/Elsa 2 version which could have this new designer?

That's OK, I will just answer with an equally annoying answer: I have no idea ;)
I honestly don't know. Could be this year, could be next year. The work itself isn't be too hard, maybe 1 week. But there's a big list of other things on my list with higher priority for the time being. Maybe if I need a break I will do it for fun. We will see.

**Comment by sfmskywalker** [comment]
Closing this issue and capturing the list with discussed designer features in #2647
